packages/next/src/server/next.ts

```diff
@@ -42,26 +43,69 @@ export type NextServerOptions = Omit<
 > &
   Partial<Pick<ServerOptions | DevServerOptions, 'conf'>>
 
-export interface RequestHandler {
-  (
-    req: IncomingMessage,
-    res: ServerResponse,
-    parsedUrl?: NextUrlWithParsedQuery | undefined
-  ): Promise<void>
-}
+export type RequestHandler = (
+  req: IncomingMessage,
+  res: ServerResponse,
+  parsedUrl?: NextUrlWithParsedQuery | undefined
+) => Promise<void>
+
+export type UpgradeHandler = (
+  req: IncomingMessage,
+  socket: Duplex,
+  head: Buffer
+) => Promise<void>
 
 const SYMBOL_LOAD_CONFIG = Symbol('next.load_config')
 
-export class NextServer {
-  private serverPromise?: Promise<Server>
-  private server?: Server
+export interface NextWrapperServer {
+  // NOTE: the methods/properties here are the public API for custom servers.
+  // Consider backwards compatibilty when changing something here!
+
+  options: NextServerOptions
+  hostname: string | undefined
+  port: number | undefined
+
+  getRequestHandler(): RequestHandler
+  prepare(serverFields?: ServerFields): Promise<void>
+  setAssetPrefix(assetPrefix: string): void
+  close(): Promise<void>
+
+  // used internally
+  getUpgradeHandler(): UpgradeHandler
+
+  // legacy methods that we left exposed in the past
+
+  logError(...args: Parameters<NextNodeServer['logError']>): void
+
+  render(
+    ...args: Parameters<NextNodeServer['render']>
+  ): ReturnType<NextNodeServer['render']>
+
+  renderToHTML(
+    ...args: Parameters<NextNodeServer['renderToHTML']>
+  ): ReturnType<NextNodeServer['renderToHTML']>
+
+  renderError(
+    ...args: Parameters<NextNodeServer['renderError']>
+  ): ReturnType<NextNodeServer['renderError']>
+
+  renderErrorToHTML(
+    ...args: Parameters<NextNodeServer['renderErrorToHTML']>
+  ): ReturnType<NextNodeServer['renderErrorToHTML']>
+
+  render404(
+    ...args: Parameters<NextNodeServer['render404']>
+  ): ReturnType<NextNodeServer['render404']>
+}
+
+/** The wrapper server used by `next start` */
+export class NextServer implements NextWrapperServer {
+  private serverPromise?: Promise<NextNodeServer>
+  private server?: NextNodeServer
   private reqHandler?: NodeRequestHandler
   private reqHandlerPromise?: Promise<NodeRequestHandler>
   private preparedAssetPrefix?: string
 
-  protected cleanupListeners: (() => Promise<void>)[] = []
```

Comment:
`cleanupListeners` is only used by `NextCustomServer`. that's one of the things that were confusing about this inheritance setup.